src/repl/session.rs

```diff
@@ -679,6 +680,47 @@ impl Session {
         output.append(&mut result);
     }
 
+    #[cfg(feature = "cli")]
+    fn reload(&mut self, output: &mut Vec<String>) {
+        self.asts.clear();
+        self.contracts.clear();
+        self.costs_reports.clear();
+        self.coverage_reports.clear();
+        self.executed.clear();
+        self.initial_contracts_analysis.clear();
+        self.interpreter =
+            ClarityInterpreter::new(
+                self.interpreter.get_tx_sender(),
+                self.settings.repl_settings.clone());
+        let contracts = self.settings.initial_contracts.clone();
+        self.settings.initial_contracts.clear();
+
+        for existing in contracts {
+            let reloaded = match fs::read_to_string(&existing.path) {
+                Ok(xs) => xs,
+                Err(_) => existing.code
```

Comment:
I think this should probably report an error.

Reply:
Hmm, perhaps something like this (typing on the fly; not really sure if it'll compile)

```rust
let reloaded = match fs::read_to_string(&existing.path) {
    Ok(code) => code,
    Err(err) => {
        output.push(
          red!(format!(
            "Error: unable to read {:?}: {}", existing.path, err)));
        return existing.code;
    }
```

Reply:
This is a side-effect inside a pattern-match, however. Not sure if that's what we really want.

Reply:
I'm okay with a side effect here. Maybe also track this so that later when you print "Ok, contracts reloaded", you can instead say something to indicate that there was a (partial) failure. Something like:

```
3 of 4 contracts successfully reloaded.
```

Note: In your code above, you'll want to change `return existing.code;` to just `existing.code`.

Reply:
Yep, it seems to work, I'm just not sure if it's idiomatic enough wrt the rest of the code base:

```rust
let reloaded;
match fs::read_to_string(&existing.path) {
    Ok(code) =>
        reloaded = code,
    Err(err) => {
        reloaded = existing.code;
        output.push(
          red!(format!(
            "Error: unable to reload {:?}: {}", existing.path, err)));
    }
}
```

Reply:
I think the more idiomatic way would be:

```rust
            let reloaded = match fs::read_to_string(&existing.path) {
                Ok(code) => code,
                Err(err) => {
                    output.push(red!(format!(
                        "Error: unable to read {}: {}",
                        existing.path, err
                    )));
                    existing.code
                }
            };
```

Reply:
> Note: In your code above, you'll want to change `return existing.code;` to just `existing.code`.

Hmm... I tried that but I got `match` arms have incompatible types. I pushed https://github.com/hirosystems/clarity-repl/pull/138/commits/272a6f8f17a6b201a0d825bb562287eb679db0b4 which seems to make `rustc` happy-er 🤔

Reply:
Make sure you get rid of the `;`. Is that the problem?

Reply:
Yes, indeed. Thank you 👍

Force pushed from [272a6f8 ](https://github.com/hirosystems/clarity-repl/commit/272a6f8f17a6b201a0d825bb562287eb679db0b4)to [44f27a2](https://github.com/hirosystems/clarity-repl/commit/44f27a2330bf096f99a2cf8dbdd4f20b9a1a7b05).

Reply:
Maybe also track this so that later when you print "Ok, contracts reloaded", you can instead say something to indicate that there was a (partial) failure. Something like:
> 
> ```
> 3 of 4 contracts successfully reloaded.
> ```

That's a good suggestion. I will look into it. 👍

Reply:
It's up to you. I would also be okay with an error reading a contract file being a catastrophic error that exits the console completely, since that likely means that the Clarinet.toml has changed, which requires re-starting anyway.

Reply:
Indeed, that sounds better (and it'll be less lines of code also).

Reply:
Done now in https://github.com/hirosystems/clarity-repl/pull/138/commits/b76aa37ce5b041bad9d779c07e5733a1463628c7.

And if/when we add support for picking up changes in Clarinet.toml (without re-starting) we can probably revisit this part so that instead of exiting it'll keep track of the contracts that could actually be reloaded.